packages/headless/src/features/insight-user-actions/insight-user-actions-preprocessing.ts

```diff
@@ -0,0 +1,339 @@
+): UserActionTimeline => {
+  const returnTimeline: UserActionTimeline = {
+    precedingSessions: [],
+    session: undefined,
+    followingSessions: [],
+  };
+
+  let currentSession: UserSession = {
+    start: actions[0].timestamp,
+    end: actions[0].timestamp,
+    actions: [],
+  };
+
+  actions.forEach((action) => {
+    if (isActionWithinSessionThreshold(action, currentSession.end)) {
+      currentSession.actions.push(action);
+      currentSession.start = action.timestamp;
+      return;
+    }
+    // We check if the case creation is within the current session or within the inactivity threshold after the end of the session.
+    const isCaseCreationIsPartOfCurrentSession = isTimestampInRange(
+      ticketCreationDate,
+      currentSession.start,
+      currentSession.end + SESSION_INACTIVITY_THRESHOLD_IN_MS
+    );
+
+    if (isCaseCreationIsPartOfCurrentSession) {
+      let caseCreationActionInserted = false;
+      const caseCreationAction: UserAction = {
+        actionType: UserActionType.TICKET_CREATION,
+        timestamp: ticketCreationDate,
+        eventData: {},
+      };
+
+      currentSession.actions.forEach((action, index) => {
+        if (caseCreationActionInserted) {
+          return;
+        }
+
+        if (
+          action.timestamp <= caseCreationAction.timestamp &&
+          !caseCreationActionInserted
+        ) {
+          currentSession.actions.splice(index, 0, caseCreationAction);
+          caseCreationActionInserted = true;
+        }
+      });
+      returnTimeline.session = currentSession;
+    } else {
+      insertSessionInTimeline(
+        currentSession,
+        ticketCreationDate,
+        returnTimeline
+      );
+    }
+
+    currentSession = {
+      start: action.timestamp,
+      end: action.timestamp,
+      actions: [action],
+    };
+  });
+
+  // We add the last session to the timeline.
+  insertSessionInTimeline(currentSession, ticketCreationDate, returnTimeline);
+
+  // If the case creation action was not part of any session added to the timeline.
+  if (returnTimeline.session === undefined) {
+    returnTimeline.session = {
+      start: ticketCreationDate,
+      end: ticketCreationDate,
+      actions: [
+        {
+          actionType: UserActionType.TICKET_CREATION,
+          timestamp: ticketCreationDate,
+          eventData: {},
+        },
+      ],
+    };
+  }
+
+  // We return the current session and the 2 preceding and following sessions in the timeline.
+  return {
+    precedingSessions: returnTimeline.precedingSessions.slice(
+      0,
+      NUMBER_OF_PRECEDING_SESSIONS
+    ),
+    session: returnTimeline.session,
+    followingSessions: returnTimeline.followingSessions.slice(
+      returnTimeline.followingSessions.length - NUMBER_OF_FOLLOWING_SESSIONS
+    ),
+  };
+};
+
+const shouldExcludeAction = (
+  action: UserAction,
+  excludedCustomActions: string[]
+): boolean => {
+  if (action.actionType === UserActionType.SEARCH && action.query === '') {
```

Comment:
Some empty search user actions doesn't even contain the field `query_expression`, so we need to cover that to make sure we exclude them as well:
```suggestion
  if (action.actionType === UserActionType.SEARCH && !action.query) {
```